### ISSUE TITLE
# dynamically changing faceMatcher for bulk photo loading

### ISSUE THREAD

**Issue by benayat** [justadudewhohacks/face-api.js#775]
I'm creating a web(react) app which **bulk loads a few tens of pictures, and then identifies and recognizes all faces together** - and sort them to albums by label\person name.

because of the obvious performance problem with bulk face recognition, I'm struggling to find the most efficient way for that -
* should I just compare the current image descriptor to all other (saved) labeled descriptors?
* or maybe is it possible to dynamically add descriptors to a big faceMatcher which will just work on every incoming picture and classify it? will it even be worth the trouble(performance-wise)? 
* any other offers will be appreciated.

**Comment by vladmandic** [comment]
if you want to be as efficient as possible, run math as fast as possible, don't look how to wrap it in nice functions that do the same thing but with additional overhead.

```js
function distance(embedding1, embedding2, order = 2) {
  if (!embedding1 || !embedding2) return 1;
  if (embedding1?.length === 0 || embedding2?.length === 0) return 1;
  if (embedding1?.length !== embedding2?.length) return 1;
  // general minkowski distance, euclidean distance is limited case (default) where order is 2
  return embedding1
    .map((val, i) => (Math.abs(embedding1[i] - embedding2[i]) ** order)) // distance squared
    .reduce((sum, now) => (sum + now), 0) // sum all distances
    ** (1 / order); // get root of
}
```

**Comment by benayat** [comment]
@vladmandic true. the next level is probably to understand how the faceMatcher work from the inside, I'm just not there yet, do to my lack of experiance\knowledge. 
from the little I can understand of your function, is the faceMatcher probability calculated by the average of euclidean distances between 68 face points and their counterparts from theother descriptor? could you please reference me to a some specific topics, so I can understand it better?
thank you for your time!

**Comment by vladmandic** [comment]
face descriptor or otherwise called face embedding is basically a tensor feature vector  
calculated on a cropped image of a face with weights trained on several versions of faces at various angles, etc.  

it's just a numerical array that describes entire face image, not related to detected face points at all  

and the length of the array is how precise it is - here it's a relatively short vector

- <https://www.tensorflow.org/hub/common_signatures/images>
- <https://medium.com/looka-engineering/how-to-visualize-feature-vectors-with-sprites-and-tensorflows-tensorboard-3950ca1fb2c7>
